datacube/storage/masking.py

```diff
@@ -109,12 +114,18 @@ def valid_data_mask(data):
     if isinstance(data, Dataset):
         return data.apply(valid_data_mask)
 
-    if isinstance(data, DataArray):
-        if 'nodata' not in data.attrs:
-            return True
-        return data != data.nodata
+    if not isinstance(data, DataArray):
+        raise TypeError('valid_data_mask not supported for type {}'.format(type(data)))
+
+    if dtype_is_float(data.dtype):
```

Comment:
use `nodata = data.attrs.get('nodata', None)`, then use local variable `nodata` instead of checking that `data.attrs['nodata']` exists and then using `data.nodata`